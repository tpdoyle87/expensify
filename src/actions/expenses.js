import database  from '../firebase/firebase';

// component calls action generator
// action generator returns object
// componnent dispatches object
// return store changes

// component calls action generator
// action generator returns a function
// componnent dispatches function (?)
// function runs ( has the ability to dispatch other actions and do whatever it wants)

// Add_expense
export const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = {description, note, amount, createdAt}
    database.ref(`users/${uid}/expenses`).push(expense).then((snapshot) => {
      dispatch(addExpense({
        id: snapshot.key,
        ...expense
      }));
    })
  }
}
// Remove_expense
export const removeExpense = ({id}) => ({
  type: "REMOVE_EXPENSE",
  id: id
})

export const startRemoveExpense = ({ id }) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({id}))
    })
  }
}
// Edit_expense
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
})

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates))
    })
  }
}

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
      const expenses = [];
      snapshot.forEach((expense) => {
        expenses.push({
          id: expense.key,
          ...expense.val()
        })
      })
      dispatch(setExpenses(expenses))
    })
  }
}
