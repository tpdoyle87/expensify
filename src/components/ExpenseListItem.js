import React from 'react'

const ExpenseListItem = ({description, amount, createdAt}) => (
  <div>
      <h3>Description: {description}</h3>
      <p>{amount}-{createdAt}</p>
  </div>
)

export default ExpenseListItem
