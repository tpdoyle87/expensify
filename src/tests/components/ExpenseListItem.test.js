import ExpenseListItem  from '../../components/ExpenseListItem'
import React from 'react'
import { shallow } from 'enzyme'
import expenses from '../fixtures/expenses'

test('it should render an expense list item', () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>)
  expect(wrapper).toMatchSnapshot();
})
