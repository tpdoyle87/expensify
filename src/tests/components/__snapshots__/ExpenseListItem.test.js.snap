// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should render an expense list item 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      Description: 
      water
    </h3>
  </Link>
  <p>
    $
    30
    -
    1
  </p>
</div>
`;
